src/blueapi/service/authentication.py

```diff
@@ -0,0 +1,218 @@
+from __future__ import annotations
+
+import base64
+import json
+import os
+import time
+from abc import ABC, abstractmethod
+from enum import Enum
+from http import HTTPStatus
+from pathlib import Path
+from typing import Any
+
+import jwt
+import requests
+
+from blueapi.config import (
+    CLIClientConfig,
+    OAuthClientConfig,
+    OAuthServerConfig,
+)
+
+
+class AuthenticationType(Enum):
+    DEVICE = "device"
+    PKCE = "pkce"
+
+
+class Authenticator:
+    def __init__(
+        self,
+        server_config: OAuthServerConfig,
+        client_config: OAuthClientConfig,
+    ):
+        self._server_config: OAuthServerConfig = server_config
+        self._client_config: OAuthClientConfig = client_config
+
+    def decode_jwt(self, token: str, verify_expiration: bool = True) -> dict[str, str]:
+        signing_key = jwt.PyJWKClient(
+            self._server_config.jwks_uri
+        ).get_signing_key_from_jwt(token)
+        decode: dict[str, str] = jwt.decode(
+            token,
+            signing_key.key,
+            algorithms=["RS256"],
+            options={"verify_exp": verify_expiration},
+            verify=True,
+            audience=self._client_config.client_audience,
+            issuer=self._server_config.issuer,
+            leeway=5,
+        )
+        return decode
+
+    def print_user_info(self, token: str) -> None:
+        decode: dict[str, str] = self.decode_jwt(token)
+        print(f'Logged in as {decode.get("name")} with fed-id {decode.get("fedid")}')
+
+
+class TokenManager(ABC):
+    @abstractmethod
+    def save_token(self, token: dict[str, Any]) -> None: ...
+    @abstractmethod
+    def load_token(token) -> dict[str, Any] | None: ...
+    @abstractmethod
+    def delete_token(self): ...
+
+
+class CliTokenManager(TokenManager):
+    def __init__(self, token_file_path: Path) -> None:
+        self._token_file_path: Path = token_file_path
+
+    def _file_path(self) -> str:
+        return os.path.expanduser(self._token_file_path)
+
+    def save_token(self, token: dict[str, Any]) -> None:
+        token_json: str = json.dumps(token)
+        token_bytes: bytes = token_json.encode("utf-8")
+        token_base64: bytes = base64.b64encode(token_bytes)
+        with open(self._file_path(), "wb") as token_file:
+            token_file.write(token_base64)
+
+    def load_token(self) -> dict[str, Any] | None:
+        file_path = self._file_path()
+        if not os.path.exists(self._file_path()):
+            return None
+        with open(file_path, "rb") as token_file:
+            token_base64: bytes = token_file.read()
+            token_bytes: bytes = base64.b64decode(token_base64)
+            token_json: str = token_bytes.decode("utf-8")
+            return json.loads(token_json)
+
+    def delete_token(self) -> None:
+        if os.path.exists(self._file_path()):
+            os.remove(self._file_path())
```

Comment:
nit: https://docs.python.org/3/library/pathlib.html#pathlib.Path.unlink
```suggestion
    def delete_token(self) -> None:
        self._file_path().unlink(missing_ok=True)
```